File: docs/csharp/scripts/error-handling.cs
#:sdk Microsoft.NET.Sdk
// Princípios: Result<T>, falhar rápido, sem exceções como controle de fluxo

Console.WriteLine(GetUser(1));
Console.WriteLine(GetUser(99));
Console.WriteLine(GetUser(0));

Result<User> GetUser(int id)
{
    if (id <= 0)
        return Result<User>.Fail(new ApiError("VALIDATION_ERROR", "id must be greater than zero"));

    var users = new List<User> { new(1, "Alice") };
    var user = users.FirstOrDefault(u => u.Id == id);

    if (user is null)
        return Result<User>.Fail(new ApiError("NOT_FOUND", $"User {id} not found"));

    return Result<User>.Ok(user);
}

record User(int Id, string Name);

record ApiError(string Code, string Message);

record Result<T>
{
    public T? Value { get; init; }
    public ApiError? Error { get; init; }
    public bool IsSuccess => Error is null;

    public static Result<T> Ok(T value) => new() { Value = value };
    public static Result<T> Fail(ApiError error) => new() { Error = error };

    public override string ToString() =>
        IsSuccess ? $"Ok({Value})" : $"Fail({Error})";
}

File: docs/typescript/scripts/error-handling.ts
// Princípios: Result pattern, typed errors, sem throws em fluxo de negócio

type Product = {
  readonly id: number;
  readonly name: string;
  readonly price: number;
};

type Result<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: string };

const saveResult = saveProduct({ id: 1, name: "Widget", price: 29.9 });
console.log(saveResult.ok ? "saved" : `error: ${saveResult.error}`);

const invalidResult = saveProduct({ id: 2, name: "", price: -1 });
console.log(invalidResult.ok ? "saved" : `error: ${invalidResult.error}`);

function saveProduct(product: Product): Result<Product> {
  const validationResult = validateProduct(product);
  if (!validationResult.ok) return validationResult;

  try {
    persistProduct(product);
    const success: Result<Product> = { ok: true, value: product };

    return success;
  } catch (error) {
    const message = error instanceof Error ? error.message : "unknown error";
    const failure: Result<Product> = { ok: false, error: `Unexpected error: ${message}` };

    return failure;
  }
}

function validateProduct(product: Product): Result<Product> {
  if (!product.name.trim()) {
    const missingName: Result<Product> = { ok: false, error: "Name is required." };
    return missingName;
  }

  if (product.price <= 0) {
    const invalidPrice: Result<Product> = { ok: false, error: "Price must be greater than zero." };
    return invalidPrice;
  }

  const valid: Result<Product> = { ok: true, value: product };

  return valid;
}

function persistProduct(product: Product): void {
  console.log(`persisting ${product.name}`);
}
